reaktive/src/commonMain/kotlin/com/badoo/reaktive/observable/FlatMap.kt

```diff
@@ -53,12 +91,56 @@ private class FlatMapObserver<in T, in R>(
 
         override fun onComplete() {
             remove(value!!)
+            queue?.poll()
             this@FlatMapObserver.onComplete()
         }
     }
 }
 
-fun <T, U, R> Observable<T>.flatMap(mapper: (T) -> Observable<U>, resultSelector: (T, U) -> R): Observable<R> =
-    flatMap { t ->
-        mapper(t).map { u -> resultSelector(t, u) }
+private class FlatMapQueue<in T : Any>(
+    limit: Int,
+    private val callback: (T) -> Unit
+) : Disposable {
+
+    private val lock = Lock()
+    private val refCounter = RefCounter(lock::destroy)
+    private val count = AtomicInt(limit)
+    private val queue = SharedQueue<T>()
+
+    private val _isDisposed = AtomicBoolean(false)
+    override val isDisposed: Boolean get() = _isDisposed.value
+
+    override fun dispose() {
+        if (_isDisposed.compareAndSet(expectedValue = false, newValue = true)) {
+            sync(queue::clear)
```

Comment:
What would happen in the case that FlatMapQueue .offer(...) just after sync(queue::clear)? It looks like it would invoke the callback with the new value despite being disposed. 
However, I don't think is a problem as the subscribe call will check the disposed state.

Reply:
The callback might be called under race condition, which will call `subscribeInner`, which subscribe the inner `Observable`. The latter will emit `Disposable`, which will be immediately disposed. Kinda same behaviour exist even without `maxConcurrency`, when upstream emits an item while the downstream is being disposed.

I updated the disposing logic inside `FlatMapQueue`, so now all resources are cleared inside the `RefCounter` block, not only the `Lock`. This won't prevent the concern, but it will be ensure there are no items left in the `Queue` at the very end.